py/path.pyi

```diff
@@ -0,0 +1,191 @@
+from typing import Any, AnyStr, Callable, ContextManager, Generic, IO, Iterable, Iterator, List, Optional, Text, Type, Union
+from typing_extensions import Final, Literal
+import os
+import sys
+
+class _FNMatcher(Generic[AnyStr]):
+    pattern: AnyStr = ...
+    def __init__(self, pattern: AnyStr) -> None: ...
+    def __call__(self, path: local) -> bool: ...
+
+class _Stat:
+    path: Final[local] = ...
+    mode: Final[int]
+    ino: Final[int]
+    dev: Final[int]
+    nlink: Final[int]
+    uid: Final[int]
+    gid: Final[int]
+    size: Final[int]
+    atime: Final[float]
+    mtime: Final[float]
+    ctime: Final[float]
+    atime_ns: Final[int]
+    mtime_ns: Final[int]
+    ctime_ns: Final[int]
+    if sys.version_info >= (3, 8) and sys.platform == "win32":
+        reparse_tag: Final[int]
+    blocks: Final[int]
+    blksize: Final[int]
+    rdev: Final[int]
+    flags: Final[int]
+    gen: Final[int]
+    birthtime: Final[int]
+    rsize: Final[int]
+    creator: Final[int]
+    type: Final[int]
+    if sys.platform != 'win32':
+        @property
+        def owner(self) -> str: ...
+        @property
+        def group(self) -> str: ...
+    def isdir(self) -> bool: ...
+    def isfile(self) -> bool: ...
+    def islink(self) -> bool: ...
+
+
+if sys.version_info >= (3, 6):
+    _PathLike = os.PathLike
+else:
+    class _PathLike(Generic[AnyStr]):
+        def __fspath__(self) -> AnyStr: ...
+_PathType = Union[bytes, Text, _PathLike[Any], local]
+
+class local:
```

Comment:
```suggestion
class local(_PathLike[str]):
```

This would allow for `pathlib.Path(py.path.local("foo"))` then, where mypy currently complains:

> Argument 1 to "Path" has incompatible type "local"; expected "Union[str, _PathLike[str]]"  [arg-type]

Reply:
Note that it should only accept Text in general (https://github.com/pytest-dev/py/pull/238).

Reply:
> This would allow for pathlib.Path(py.path.local("foo")) then, where mypy currently complains:

The situation is odd. If we also do it for the real `local` (in the `.py`, not just in the `.pyi`) then that would remove any doubt from my mind.

For now a workaround for `pathlib.Path(py.path.local("foo"))` is `pathlib.Path(str(py.path.local("foo")))`. This is also be portable to Python<=3.5.

Reply:
Yeah, the workaround is known, but currently mypy complains about something that works (with py36+).
We can leave it out for now if unclear, but should probably make it more strict to not accepts bytes?

Reply:
> We can leave it out for now if unclear

If we make the real `local` derive from `os.PathLike`, then it would be good to add it in the stubs too. But if not, then the stub would be sort-of lying. Although `os.PathLike` has a hack where `isinstance(foo, os.PathLike)` returns true even if `foo` just implements `__fspath__`, without deriving. So I'm not really sure.

> but should probably make it more strict to not accepts bytes?

While it seems `bytes` support is broken, I'm sure there's someone someplace that uses it. For example https://github.com/pytest-dev/py/issues/187#issuecomment-400420032. So again, not sure :)

Reply:
An article on how pathlib (does not) handle this: https://jodal.no/2019/12/10/pathlib-and-paths-with-arbitrary-bytes/